### PR TITLE
Feature/cornflow core

### PR DESCRIPTION
**Pull request by ggsdc**
This library should have the base data models (abstracts) the main logic for the authentication (simple jwt token), the meta resource (to create endpoints) and some base schemas (for the base models and for the queries for example).

Maybe this library could also host the code that @marioncottard has developed to create API RESTS from a json schema, as it could use these base classes to easily set up either just the models / schemas / endpoints or the full flask app based on the rest of the base components such as the authentication.

### CODE REVIEW

File: libs/core/cornflow_core/models/user.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        Updates the user information in the database
+
+        :param dict data: the data to update the user
+        """
+        # First we create the hash of the new password and then we update the object
+        new_password = data.get("password")
+        if new_password:
+            new_password = self.__generate_hash(new_password)
+            data["password"] = new_password
+        super().update(data)
```
Comment (marioncottard) [inline libs/core/cornflow_core/models/user.py:60]: Maybe you should also check the password pattern here ?

**Review by marioncottard — COMMENTED** [review]
A few things:

File: libs/core/cornflow_core/resources/meta_resource.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        :rtype: bool
+        """
+        return self.get_user().is_admin()
+
+    def is_service_user(self):
+        """
+        :return: if user is service user
+        :rtype: bool
+        """
+        return self.get_user().is_service_user()
```
[the remark this answers is not in the document]
Comment (ggsdc) [reply libs/core/cornflow_core/resources/meta_resource.py:111]: That could be a good solution to this, that way we can force to have them set up

File: libs/core/cornflow_core/models/user.py (hunk 1 above)
Comment (ggsdc) [reply libs/core/cornflow_core/models/user.py:60]: Now it get checked on the endpoint logic, not on the user model if I remember corrrectly

**Comment by ggsdc** [comment]
This should cover all that has to be set up on this library, except your changes for the commands to generate new REST APIs.

A lot of documentation is missing (docstrings mainly) and tests. Haven't thought of how to test it without testing it directly on cornflow-server, which should not be the best case as potentially this could have things in the future that are used by other apps but not by cornflow-server

File: libs/core/cornflow_core/models/role.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    def __repr__(self):
+        return f"<Role {self.name}>"
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class UserRoleBaseModel(TraceAttributesModel):
+    __abstract__ = True
+    id = database.Column(database.Integer, primary_key=True, autoincrement=True)
```
[the remark this answers is not in the document]
Comment (marioncottard) [reply libs/core/cornflow_core/models/role.py]: Oh, okay, I didn't know that !

**Comment by ggsdc** [comment]
@marioncottard can you review this PR? If it all right I will make the final changes to the cornflow_tools branch to join it with this

**Comment by marioncottard** [comment]
Okay, I'll review it now

**Review by marioncottard — APPROVED** [review]
Looks good to me!

**Pull request merged**